src/NServiceBus.Azure/Subscriptions/Azure/AzureSubscriptionStorage.cs

```diff
@@ -76,28 +77,40 @@ void ISubscriptionStorage.Unsubscribe(Address address, IEnumerable<MessageType>
 
         IEnumerable<Address> ISubscriptionStorage.GetSubscriberAddressesForMessage(IEnumerable<MessageType> messageTypes)
         {
-            var subscribers = new List<Address>();
+            var subscribers = new HashSet<Address>();
 
             using (var context = new SubscriptionServiceContext(client))
             {
                 foreach (var messageType in messageTypes)
                 {
-                    var type = messageType;
+                    var name = messageType.TypeName;
+                    var lowerBound = name;
+                    var upperBound = GetUpperBound(name);
+
                     var query = from s in context.Subscriptions
-                                where s.PartitionKey == type.ToString() 
-                                select s;
+                        where s.PartitionKey.CompareTo(lowerBound) >= 0 &&
+                              s.PartitionKey.CompareTo(upperBound) < 0
```

Comment:
remark: there is a potential table scan here, but as this table is usually small I don't think it is an issue